### PR TITLE
Initial implementation of ObservableString.

### PR DESCRIPTION
**Pull request by ian-r-rose**
Opening this for discussion.

We talked [earlier](https://www.youtube.com/watch?v=1l2J7EjfsgA) about the implementation of an `ObservableString` type which allows you to subscribe to changes in the text of the string. I would find this useful for realtime collaboration (at least in my current thinking). This PR has a very basic implementation based on that of the `ObservableVector`. Questions I have at the moment:
- How granular should the changes be? Is it worthwhile to actually do some parsing of string diffs?
- Related to the above: are performance characteristics of Javascript engines for string parsing well understood enough to inform the design of the type? Is anyone aware of any good reference for that?
- Is this a useful thing elsewhere in the codebase?

### CODE REVIEW

**Comment by blink1073** [comment]
That could be used for example to set a finite range in a text editor which could be done more  efficiently.

**Comment by blink1073** [comment]
This could be used for example in the abstract editor interface for the `value` field, allowing you to set a range of the value.

**Comment by ian-r-rose** [comment]
Yes, this is just a bare minimum implementation. Some more granular methods would be useful, but before implementing them I was hoping to get a better understanding of string parsing performance of Javascript engines. I could imagine that a naive implementation would have worse performance than just simple resetting or concatenation (as @sccolbert implied).

But yes, I could see this being useful in the abstract editor interface.

**Comment by ian-r-rose** [comment]
Okay, I have reverted the changes to the default document model for now (pending some refactoring as per an in-person discussion).

File: src/common/observablestring.ts [hunk 1]
```diff
@@ -0,0 +1,215 @@
...
+  changed: ISignal<IObservableString, ObservableString.IChangedArgs>;
+
+  /**
+   * The value of the string.
+   */
+  text : string;
+
+  /**
+   * Insert a string, where `index` is the starting
+   * index, and `text` is the string to insert.
```
Comment (blink1073) [inline src/common/observablestring.ts]: Can we modify these to be more like:

```
/**
 * Insert a string.
 *
 * @param index - The starting index.
 *
 * @param text - The string to insert.
 */
```

File: src/common/observablestring.ts [hunk 2]
```diff
@@ -0,0 +1,215 @@
...
+  dispose(): void;
+}
+
+/**
+ * A concrete implementation of [[IObservableString]] 
+ */
+export
+class ObservableString implements IObservableString {
+
+  constructor(initialText: string = '') {
```
Comment (blink1073) [inline src/common/observablestring.ts]: Needs a doc: "Construct a new observable string"

File: src/common/observablestring.ts [hunk 3]
```diff
@@ -0,0 +1,215 @@
...
+
+  /**
+   * Get the value of the string.
+   */
+  get text(): string {
+    return this._text;
+  }
+
+  /**
+   * Insert a string, where `index` is the starting
```
Comment (blink1073) [inline src/common/observablestring.ts]: These will need to be updated to reflect the interface docs.

File: src/common/observablestring.ts [hunk 4]
```diff
@@ -0,0 +1,215 @@
...
+  private _text = '';
+  private _isDisposed : boolean = false;
+}
+
+/**
+ * The namespace for `ObservableVector` class statics.
+ */
+export
+namespace ObservableString {
+
```
Comment (blink1073) [inline src/common/observablestring.ts]: Extraneous extra line.

File: src/common/observablestring.ts [hunk 5]
```diff
@@ -0,0 +1,215 @@
...
+     */
+    start: number;
+
+    /**
+     * The end index of the change
+     */
+    end: number;
+
+    /**
+     * The value of the change.
```
Comment (blink1073) [inline src/common/observablestring.ts]: Can you move everything after the first sentence to notes?

```
* #### Notes
* If ...
```

**Comment by ian-r-rose** [comment]
Okay, comments addressed.

File: src/common/observablestring.ts [hunk 6]
```diff
@@ -0,0 +1,233 @@
...
+      end: end,
+      value: oldValue
+    });
+  }
+
+  /**
+   * Set the ObservableString to an empty string.
+   */
+  clear(): void {
+    this._text = '';
```
Comment (blink1073) [inline src/common/observablestring.ts]: Implement as `this.text = ''`?

File: src/common/observablestring.ts [hunk 7]
```diff
@@ -0,0 +1,233 @@
...
+   */
+  dispose(): void;
+}
+
+/**
+ * A concrete implementation of [[IObservableString]] 
+ */
+export
+class ObservableString implements IObservableString {
+
```
Comment (blink1073) [inline src/common/observablestring.ts]: Extraneous line.

File: src/common/observablestring.ts [hunk 8]
```diff
@@ -0,0 +1,233 @@
...
+interface IObservableString extends IDisposable {
+  /**
+   * A signal emitted when the string has changed.
+   */
+  changed: ISignal<IObservableString, ObservableString.IChangedArgs>;
+
+  /**
+   * The value of the string.
+   */
+  text : string;
```
Comment (blink1073) [inline src/common/observablestring.ts]: Extraneous space after `text`

File: src/common/observablestring.ts [hunk 9]
```diff
@@ -0,0 +1,232 @@
...
+/**
+ * A concrete implementation of [[IObservableString]] 
+ */
+export
+class ObservableString implements IObservableString {
+  /**
+   * Construct a new observable string.
+   */
+  constructor(initialText: string = '') {
+    this.text = initialText;
```
Comment (blink1073) [inline src/common/observablestring.ts]: No need to emit a change on startup, this should be `this._text =`

File: src/common/observablestring.ts [hunk 10]
```diff
@@ -0,0 +1,232 @@
...
+      value: oldValue
+    });
+  }
+
+  /**
+   * Set the ObservableString to an empty string.
+   */
+  clear(): void {
+    this.text = '';
+    this.changed.emit({
```
Comment (blink1073) [inline src/common/observablestring.ts]: No need to emit the event, it will be taken care of in the setter.

Comment (ian-r-rose) [inline src/common/observablestring.ts]: Whoops, right you are.

**Comment by ian-r-rose** [comment]
Whoops, right you are/

**Comment by blink1073** [comment]
This looks good @ian-r-rose, is it ready?

**Comment by ian-r-rose** [comment]
Yes, should be ready.

**Comment by blink1073** [comment]
Great, thank you!

**Pull request merged**